src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -4723,3 +4735,249 @@ void C2_MacroAssembler::vector_reverse_byte(BasicType bt, XMMRegister dst, XMMRe
+                                                        XMMRegister xtmp1, XMMRegister xtmp2, XMMRegister xtmp3,
+                                                        KRegister ktmp, Register rtmp, bool merge, int vec_enc) {
+  assert(is_integral_type(bt), "");
+  assert(VM_Version::supports_avx512vl() || vec_enc == Assembler::AVX_512bit, "");
+  assert(VM_Version::supports_avx512cd(), "");
+  switch(bt) {
+    case T_LONG:
+      evplzcntq(dst, ktmp, src, merge, vec_enc);
+      break;
+    case T_INT:
+      evplzcntd(dst, ktmp, src, merge, vec_enc);
+      break;
+    case T_SHORT:
+      vpternlogd(xtmp1, 0xff, xtmp1, xtmp1, vec_enc);
+      vpunpcklwd(xtmp2, xtmp1, src, vec_enc);
+      evplzcntd(xtmp2, k0, xtmp2, merge, vec_enc);
+      vpunpckhwd(dst, xtmp1, src, vec_enc);
+      evplzcntd(dst, k0, dst, merge, vec_enc);
+      vpackusdw(dst, xtmp2, dst, vec_enc);
+      break;
+    case T_BYTE:
+      // T1 = Compute leading zero counts of 4 LSB bits of each byte by
+      // accessing the lookup table.
+      // T2 = Compute leading zero counts of 4 MSB bits of each byte by
+      // accessing the lookup table.
+      // Add T1 to T2 if 4 MSB bits of byte are all zeros.
+      assert(VM_Version::supports_avx512bw(), "");
+      evmovdquq(xtmp1, ExternalAddress(StubRoutines::x86::vector_count_leading_zeros_lut()), vec_enc, rtmp);
+      movl(rtmp, 0x0F0F0F0F);
+      evpbroadcastd(dst, rtmp, vec_enc);
+      vpand(xtmp2, dst, src, vec_enc);
+      vpshufb(xtmp2, xtmp1, xtmp2, vec_enc);
+      vpsrlw(xtmp3, src, 4, vec_enc);
+      vpand(xtmp3, dst, xtmp3, vec_enc);
+      vpshufb(dst, xtmp1, xtmp3, vec_enc);
+      vpxor(xtmp1, xtmp1, xtmp1, vec_enc);
+      evpcmpeqb(ktmp, xtmp1, xtmp3, vec_enc);
+      evpaddb(dst, ktmp, dst, xtmp2, true, vec_enc);
+      break;
+    default:
+      ShouldNotReachHere();
+  }
+}
+
+void C2_MacroAssembler::vector_count_leading_zeros_byte_avx(XMMRegister dst, XMMRegister src, XMMRegister xtmp1,
+                                                            XMMRegister xtmp2, XMMRegister xtmp3, Register rtmp, int vec_enc) {
+  vmovdqu(xtmp1, ExternalAddress(StubRoutines::x86::vector_count_leading_zeros_lut()));
+  movl(rtmp, 0x0F0F0F0F);
+  movdl(xtmp2, rtmp);
+  vpbroadcastd(xtmp2, xtmp2, vec_enc);
+  // T1 = Compute leading zero counts of 4 LSB bits of each byte by
+  // accessing the lookup table.
+  vpand(dst, xtmp2, src, vec_enc);
+  vpshufb(dst, xtmp1, dst, vec_enc);
+  // T2 = Compute leading zero counts of 4 MSB bits of each byte by
+  // accessing the lookup table.
+  vpsrlw(xtmp3, src, 4, vec_enc);
+  vpand(xtmp3, xtmp2, xtmp3, vec_enc);
+  vpshufb(xtmp2, xtmp1, xtmp3, vec_enc);
+  // Add T1 to T2 if 4 MSB bits of byte are all zeros.
+  vpxor(xtmp1, xtmp1, xtmp1, vec_enc);
+  vpcmpeqb(xtmp3, xtmp1, xtmp3, vec_enc);
+  vpaddb(dst, dst, xtmp2, vec_enc);
+  vpblendvb(dst, xtmp2, dst, xtmp3, vec_enc);
```

Comment:
vector_count_leading_zeros_byte_avx can be implemented without xmpt3 as follows:
 // Nibble clz table in xtmp1
 vmovdqu(xtmp1, ExternalAddress(StubRoutines::x86::vector_count_leading_zeros_lut()), rtmp);
 
// Nibble mask in xtmp2
 vbroadcast(T_INT, xtmp2, 0x0F0F0F0F, rtmp, vec_enc);

  // dst = Compute leading zero counts of 4 MSB bits of each byte by
  // accessing the lookup table
  vpsrlw(dst, src, 4, vec_enc);
  vpand(dst, dst, xtmp2, vec_enc);
  vpshufb(dst, xtmp1, dst, vec_enc);

  // xtmp2 = Compute leading zero counts of 4 LSB bits of each byte by
  // accessing the lookup table.
  vpand(xtmp2, xtmp2, src, vec_enc);
  vpshufb(xtmp2, xtmp1, xtmp2, vec_enc);
  
  // Add xtmp2 to dst if 4 MSB bits of byte are all zeros i.e. if the dst had clz of 0
  vpbroadcastb(xtmp1, xtmp1, vec_enc); 
  vpcmpeqb(xtmp1, xtmp1, dst, vec_enc);
  vpaddb(xtmp2, xtmp2, dst, vec_enc);
  vpblendvb(dst, dst, xtmp2, xtmp1, vec_enc);

Reply:
Thanks, Sandhya,  you have introduced additional broadcast instruction to save a temporary.  Spill to memory may be costly too but adding an explicit instruction in anticipation of a spill may hit the performance. Also we will not be able to further merge this instruction pattern even after removing additional temporary. 

// Add xtmp2 to dst if 4 MSB bits of byte are all zeros i.e. if the dst had clz of 0
We add LZ count of 4MSB bits if they are all zeros i.e. clz of 4.